src/Util/MigrationRollbackBatch.php

```diff
@@ -0,0 +1,288 @@
+
+/**
+ * Class responsible for rolling back a migration in batches.
+ */
+class MigrationRollbackBatch extends MigrateExecutable {
+
+  use DependencySerializationTrait {
+    __sleep as traitSleep;
+    __wakeup as traitWakeup;
+  }
+
+  /**
+   * Messenger service to display messages.
+   *
+   * @var Drupal\Core\Messenger\MessengerInterface
+   */
+  protected MessengerInterface $messenger;
+
+  /**
+   * Iterates through migration rows.
+   *
+   * @var Drupal\dgi_migrate\MigrationIteratorMigrationIterator
+   */
+  protected MigrationIterator $iterator;
+
+  /**
+   * Options.
+   */
+  private array $options;
+
+  /**
+   * Stores the migration rows
+   *
+   * @var QueueInterface
+   */
+  protected QueueInterface $queue;
+
+  /**
+   * Batch context
+   *
+   * @var array
+   */
+  protected array $context;
+
+  /**
+   * Constructs a new MigrationRollbackBatch instance.
+   *
+   * @param \MigrationInterface $migration
+   *   The migration interface.
+   * @param \MessengerInterface $messenger
+   *   The messenger service.
+   * @param array $options
+   *   An array of options.
+   */
+  public function __construct(
+    MigrationInterface $migration,
+    MessengerInterface $messenger,
+    array $options,
+  ) {
+    parent::__construct($migration, new MigrateMessage(), $options);
+    $this->options = $options;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * Prepare a batch array for execution for the given migration.
+   *
+   * @return array
+   *   A batch array with operations and the like.
+   *
+   * @throws \Exception
+   *   If the migration could not be enqueued successfully.
+   */
+  public function prepareBatch(): array {
+    $id_map = $this->getIdMap();
+
+    if (isset($this->options['checkStatus']) && $this->options['checkStatus'] == 1) {
+      // Filter for failed rows.
+      $failedStatuses = StatusFilter::mapStatuses('failed');
+      $filteredIdMap = new StatusFilter($id_map, $failedStatuses);
+      $this->iterator = new MigrationIterator($filteredIdMap, 'currentDestination');
+    }
+    else {
+      $this->iterator = new MigrationIterator($id_map, 'currentDestination');
+    }
+
+    $result = $this->enqueue();
+
+    if ($result === MigrationInterface::RESULT_COMPLETED) {
+      return [
+        'title' => $this->t('Rolling back migration: @migration', [
+          '@migration' => $this->migration->id(),
+        ]),
+        'operations' => [
+          [[$this, 'processBatch'], []],
+        ],
+        'finished' => [$this, 'finishBatch'],
+      ];
+    }
+    else {
+      throw new \Exception('Migration failed.');
+    }
+  }
+
+  /**
+   * Adds the migration rows to a queue
+   *
+   * @return int
+   */
+  private function enqueue(): int {
+    $this->iterator->rewind();
+
+    $this->queue = $this->getQueue();
+    $this->queue->deleteQueue();
+
+    while ($this->iterator->valid()) {
+      $this->queue->createItem($this->iterator->current());
+      $this->iterator->next();
+    }
```

Comment:
For the "more elaborate" data in https://github.com/discoverygarden/islandora_spreadsheet_ingest/pull/108/files#r1629714922 , thinking something like:

```suggestion
    foreach ($this->iterator as $row) {
      $this->queue->createItem([
        'source' => $this->iterator->getInnerIterator()->currentSource(),
        'destination' => $row,
      ]);
    }
```

(is `foreach` also more readable?)

and of course, carrying through the references to the new array keys where the data is read.

Reply:
Another note, but `foreach` would also handle the rewinding the iterator.